### PR TITLE
WIP: Dependency manager

### PR DESCRIPTION
**Pull request by Glavin001**
See https://github.com/Unibeautify/unibeautify/issues/58

Example: https://github.com/Unibeautify/beautifier-php-cs-fixer

### CODE REVIEW

**Comment by Glavin001** [comment]
Do not merge. Still lots to do!

Please feel free to make comments. This is a critical feature and I want it built right.

**Comment by Glavin001** [comment]
@szeck87 : Can now see Coverage page by adding `/coverage` to Netlify preview link: https://5abc7023df99537901e29491--unibeautify-api.netlify.com/coverage/

File: src/DependencyManager/Dependency.ts [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+export abstract class Dependency {
+  private _isInstalled: boolean = false;
+  private _version?: Version;
+  private _errors: Error[] = [];
+
+  constructor(protected options: DependencyOptions) {}
+
+  public load(): Promise<boolean> {
+    if (this.isInstalled) {
+      return Promise.resolve(this.isInstalled);
```
Comment (szeck87) [inline src/DependencyManager/Dependency.ts]: This needs code coverage.

File: src/DependencyManager/Dependency.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+  }
+
+  private extractWithPatterns(
+    text: string,
+    patterns: (string | RegExp)[]
+  ): string | undefined {
+    return patterns.reduce(
+      (extracted: string | undefined, pattern: string | RegExp) => {
+        if (extracted) {
+          return extracted;
```
Comment (szeck87) [inline src/DependencyManager/Dependency.ts]: This needs code coverage.

File: src/DependencyManager/ExecutableDependency.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  }): Promise<RunResponse> {
+    return new Promise((resolve, reject) => {
+      const cmd = spawn(exe, args, options);
+      let stdout = "";
+      let stderr = "";
+      cmd.stdout.on("data", data => {
+        return (stdout += data);
+      });
+      cmd.stderr.on("data", data => {
+        return (stderr += data);
```
Comment (szeck87) [inline src/DependencyManager/ExecutableDependency.ts]: This needs code coverage.

File: src/DependencyManager/ExecutableDependency.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+      });
+      cmd.on("close", exitCode => {
+        return resolve({ exitCode, stdout, stderr });
+      });
+      cmd.on("error", err => {
+        return reject(err);
+      });
+      if (stdin) {
+        cmd.stdin.end(stdin);
```
Comment (szeck87) [inline src/DependencyManager/ExecutableDependency.ts]: This needs code coverage.

**Comment by szeck87** [comment]
And if we want to hit 100% code coverage, this line too: 
 https://github.com/Unibeautify/unibeautify/blob/dependency-manager/src/InlineFlagManager.ts#L31

**Pull request merged**